src/jit/lower.cpp

```diff
@@ -3761,6 +3843,197 @@ void Lowering::LowerAdd(GenTreePtr* pTree, Compiler::fgWalkData* data)
+// LowerUnsignedDivOrMod: transform GT_UDIV/GT_UMOD nodes with a const power of 2
+// divisor into GT_RSZ/GT_AND nodes.
+//
+// Arguments:
+//    tree:   pointer to the GT_UDIV/GT_UMOD node to be lowered
+
+void Lowering::LowerUnsignedDivOrMod(GenTree* tree)
+{
+    assert(tree->OperGet() == GT_UDIV || tree->OperGet() == GT_UMOD);
+
+    GenTree* divisor = tree->gtGetOp2();
+
+    if (divisor->IsCnsIntOrI())
+    {
+        size_t divisorValue = static_cast<size_t>(divisor->gtIntCon.IconValue());
+
+        if (isPow2(divisorValue))
+        {
+            genTreeOps newOper;
+
+            if (tree->OperGet() == GT_UDIV)
+            {
+                newOper = GT_RSZ;
+                divisorValue = genLog2(divisorValue);
+            }
+            else
+            {
+                newOper = GT_AND;
+                divisorValue -= 1;
+            }
+
+            tree->SetOper(newOper);
+            divisor->gtIntCon.SetIconValue(divisorValue);
+        }
+    }
+}
+
+//------------------------------------------------------------------------
+// LowerSignedDivOrMod: transform integer GT_DIV/GT_MOD nodes with a power of 2 
+// const divisor into equivalent but faster sequences.
+//
+// Arguments:
+//    pTree:   pointer to the parent node's link to the node we care about
+
+void Lowering::LowerSignedDivOrMod(GenTreePtr* ppTree, Compiler::fgWalkData* data)
+{
+    GenTree* divMod = *ppTree;
+    assert(divMod->OperGet() == GT_DIV || divMod->OperGet() == GT_MOD);
+    GenTree* divisor = divMod->gtGetOp2();
+
+    if (divisor->IsCnsIntOrI())
+    {
+        const var_types type = divMod->TypeGet();
+        assert(type == TYP_INT || type == TYP_LONG);
+
+        GenTree* dividend = divMod->gtGetOp1();
+
+        if (dividend->IsCnsIntOrI())
+        {
+            // We shouldn't see a divmod with constant operands here but if we do then it's likely 
+            // because optimizations are disabled or it's a case that's supposed to throw an exception. 
+            // Don't optimize this.
+            return;
+        }
+
+        ssize_t divisorValue = divisor->gtIntCon.IconValue();
+
+        if (divisorValue == -1)
+        {
+            // x / -1 can't be optimized because INT_MIN / -1 is required to throw an exception.
+         
+            // x % -1 is always 0 and the IL spec says that the rem instruction "can" throw an exception if x is
+            // the minimum representable integer. However, the C# spec says that an exception "is" thrown in this
+            // case so optimizing this case would break C# code.
+
+            // A runtime check could be used to handle this case but it's probably too rare to matter.
+            return;
+        }
+        
+        bool isDiv = divMod->OperGet() == GT_DIV;
+
+        if (isDiv)
+        {
+            if ((type == TYP_INT && divisorValue == INT_MIN) ||
+                (type == TYP_LONG && divisorValue == INT64_MIN))
```

Comment:
I see from below that you have tests that exercise this case, but I wonder if this will require special support for 32-bit targets, since the MSIL relational operators always return an int, and I suspect the JIT assumes the same for the GT_EQ and similar operators.